### PR TITLE
SALTO-4878: OAuth for Okta - part 1

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Change adapter code so it will be able to get `clientId`, `clientSecret`, `accessToken` and `refreshToken`, as credentials, and refresh access token a create connection with it.
This code is enough to start implementing OAUTH in saas, but we will need to make adjustment to support oauth flow for OSS.

---

_Additional context for reviewer_

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/okta-adapter/src/auth.ts [hunk 1]
```diff
@@ -22,6 +22,14 @@ export type AccessTokenCredentials = {
   token: string
 }
 
+export type OAuthAccessTokenCredentials = {
+  baseUrl: string
+  accessToken: string
+  refreshToken?: string
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/auth.ts:28]: currently optional as I thought that for OSS flow we won't need refresh token

File: packages/okta-adapter/src/adapter_creator.ts [hunk 2]
```diff
@@ -48,15 +48,20 @@ const {
...
     throw new Error(`'${hostName}' is not a valid okta account url`)
   }
-  return {
-    baseUrl,
-    token,
-  }
+  return authType === 'oauth'
+    ? {
+      baseUrl,
+      accessToken: config.value.accessToken,
```
Comment (dantal4) [inline packages/okta-adapter/src/adapter_creator.ts:59]: can be removed?

File: packages/okta-adapter/src/adapter_creator.ts [hunk 3]
```diff
@@ -48,15 +48,20 @@ const {
...
+      clientId: config.value.clientId,
+      clientSecret: config.value.clientSecret,
+      refreshToken: config.value.refreshToken,
```
Comment (dantal4) [inline packages/okta-adapter/src/adapter_creator.ts:62]: maybe use a schema guard to verify `config.value` ?

File: packages/okta-adapter/src/auth.ts [hunk 4]
```diff
@@ -36,4 +44,36 @@ export const accessTokenCredentialsType = createMatchingObjectType<AccessTokenCr
   },
 })
 
-export type Credentials = AccessTokenCredentials
+export const OAuthAccessTokenCredentialsType = createMatchingObjectType<
```
Comment (dantal4) [inline packages/okta-adapter/src/auth.ts:47]: nit
```suggestion
export const oauthAccessTokenCredentialsType = createMatchingObjectType<
```

File: packages/okta-adapter/src/client/connection.ts [hunk 5]
```diff
@@ -60,21 +60,44 @@ export const validateCredentials = async (_creds: {
   }
 }
 
-const accessTokenAuthParamsFunc = (
+const APITokenAuthParamsFunc = (
```
Comment (dantal4) [inline packages/okta-adapter/src/client/connection.ts:63]: nit
```suggestion
const apiTokenAuthParamsFunc = (
```

File: packages/okta-adapter/src/client/connection.ts [hunk 6]
```diff
@@ -60,21 +60,44 @@ export const validateCredentials = async (_creds: {
...
   { token }: AccessTokenCredentials
 ): clientUtils.AuthParams => ({
   headers: {
     Authorization: `SSWS ${token}`,
   },
 })
 
+const OAuthAuthParamsFunc = async (
```
Comment (dantal4) [inline packages/okta-adapter/src/client/connection.ts:71]: nit
```suggestion
const oauthAuthParamsFunc = async (
```

File: packages/okta-adapter/src/auth.ts (hunk 4 above)
Comment (dantal4) [inline packages/okta-adapter/src/auth.ts:47]: I think that this should be used in the adapter creator?
```javascript
authenticationMethods: {
  basic: {...},
  oauth: {
    credentialsType: oauthAccessTokenCredentialsType,
  },
}
```

**Review by dantal4 — COMMENTED** [review]
Great work!

File: packages/adapter-components/src/auth/oauth.ts [hunk 7]
```diff
@@ -81,3 +82,51 @@ export const oauthClientCredentialsBearerToken = async ({
...
+  baseURL,
+  clientId,
+  clientSecret,
+  refreshToken,
+  retryOptions,
+}: {
+  endpoint: string
+  baseURL: string
+  clientId: string
+  clientSecret: string
```
[the remark this answers is not in the document]
Comment (shir-reifenberg) [reply packages/adapter-components/src/auth/oauth.ts:100]: Talked IRL, but we'll ask users to provide those as it changes

File: packages/okta-adapter/src/adapter_creator.ts (hunk 2 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/adapter_creator.ts:59]: yes 🙏

File: packages/okta-adapter/src/auth.ts (hunk 4 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/auth.ts:47]: I think it's needed for the open source authentication flow but I didn't see the saas code uses this as well

**Review by dantal4 — APPROVED** [review]
💯

**Pull request merged**